std/algorithm/iteration.d

```diff
@@ -2945,6 +2949,389 @@ template fold(fun...) if (fun.length >= 1)
+the same value as $(D fold!(fun)(seed, range)).
+The one-argument version $(D cumulativeFold!(fun)(range)) works similarly, but
+it returns the first element unchanged and uses it as seed for the next
+elements.
+This function is also known as `partial_sum`, `accumulate`, `prefix_sum`,
+`scan`, or `cumulative_sum`.
+
+Returns:
+    a range containg the consecutive reduced values. If there is more than one
+    fun, the element type will be $(REF Tuple, std, typecons) containing one
+    element for each fun.
+
+See_Also:
+    $(WEB en.wikipedia.org/wiki/Prefix_sum, Prefix Sum)
++/
+template cumulativeFold(fun...)
+if (fun.length >= 1)
+{
+    import std.meta : staticMap;
+    alias binfuns = staticMap!(binaryFun, fun);
+
+    /++
+    No-seed version. The first element of $(D r) is used as the seed's value.
+    For each function $(D f) in $(D fun), the corresponding seed type $(D S) is
+    $(D Unqual!(typeof(f(e, e)))), where $(D e) is an element of $(D r): $(D
+    ElementType!R).
+    Once S has been determined, then $(D S s = e;) and $(D s = f(s, e);) must
+    both be legal.
+    Params:
+        fun = one or more functions
+        range = an input range as defined by $(D isInputRange)
+    Returns:
+        a range containing the consecutive reduced values.
+    +/
+    auto cumulativeFold(R)(R range)
+    if (isInputRange!(Unqual!R))
+    {
+        return cumulativeFoldImpl(range);
+    }
+
+    /++
+    Seed version. The seed should be a single value if $(D fun) is a
+    single function. If $(D fun) is multiple functions, then $(D seed)
+    should be a $(XREF typecons,Tuple), with one field per function in $(D f).
+    For convenience, if the seed is const, or has qualified fields, then
+    $(D cumulativeFold) will operate on an unqualified copy. If this happens
+    then the returned type will not perfectly match $(D S).
+    Params:
+        fun = one or more functions
+        range = an input range as defined by $(D isInputRange)
+        seed = the initial value of the accumulator
+    Returns:
+        a range containing the consecutive reduced values.
+    +/
+    auto cumulativeFold(R, S)(R range, S seed)
+    if (isInputRange!(Unqual!R))
+    {
+        static if (fun.length == 1)
+            return cumulativeFoldImpl(range, seed);
+        else
+            return cumulativeFoldImpl(range, seed.expand);
+    }
+
+    private auto cumulativeFoldImpl(R, Args...)(R range, ref Args args)
+    {
+        import std.algorithm.internal : algoFormat;
+
+        static assert(Args.length == 0 || Args.length == fun.length,
+            algoFormat("Seed %s does not have the correct amount of fields (should be %s)",
+                Args.stringof, fun.length));
+
+        static if (args.length)
+            alias State = staticMap!(Unqual, Args);
+        else
+            alias State = staticMap!(ReduceSeedType!(ElementType!R), binfuns);
+
+        foreach (i, f; binfuns)
```

Comment:
I'm not sure that I agree that `foreach` shouldn't be used in generic code. Sometimes, it's exactly what's appropriate. It's just that there's often a better way to do it that doesn't involve `foreach`, and you have to be careful with the element type, since `foreach` will infer strings to have their code unit as their element type and not dchar.

Regardless, in this case, it looks like this is a static foreach (binfuns is the result of `staticMap`), so it pretty much has to be done either with a `foreach` or a recursive template. The `foreach` is likely to be cleaner and more efficient, particularly when this is inside of a function rather than something that's part of the public API.